File: ad/ad_assignments/a9.hs
import Data.List


nil = (-1, -1, -1)
nil_elem = -1

bestInvestment :: [Int] -> (Int, Int, Int)
bestInvestment [] = nil
bestInvestment [x] = nil
bestInvestment kurs = bestInvest minimum maximum
	where
		minimum = minimumList kurs 
		maximum = maximumList kurs


bestInvest :: [Int] -> [Int] -> (Int, Int, Int)
bestInvest [] _ = nil
bestInvest kmin kmax 
			| max_difference > 0 = (minimum_idx, maximum_idx, max_difference) 
			| otherwise = nil
			where
				kzip = (zip kmin kmax)
				minimum_idx = minimumIndex kzip
				difference = kzip !! minimum_idx
				max_difference = calcDiff difference
				maximum_idx = maximumIndex kzip minimum_idx difference
			

calcDiff :: (Int, Int) -> Int
calcDiff x = (snd x) - (fst x)


minimumIndex :: [(Int, Int)] -> Int
minimumIndex [] = nil_elem
minimumIndex (x:xs) = midx' x (x:xs) 0 0
		where
			midx' _ [] idx max_idx = max_idx
			midx' y (z:zs) idx max_idx
				| calcDiff y >= calcDiff z = midx' y zs (idx + 1) max_idx
				| otherwise =  midx' z zs (idx + 1) idx

maximumIndex :: [(Int, Int)] -> Int -> (Int, Int) -> Int
maximumIndex [] _ _ = nil_elem
maximumIndex xs min_idx max_diff = max_offset + min_idx - 1
		where
			k_from_min = drop min_idx xs
			max_offset = length $ takeWhile (==max_diff) k_from_min


minimumList :: [Int] -> [Int]
minimumList [] = []
minimumList (x:xs) = mlist' min x (x:xs)


maximumList :: [Int] -> [Int]
maximumList [] = []
maximumList xs = reverse $ mlist' max (last xs) (reverse xs)


mlist' :: (Int -> Int -> Int) -> Int -> [Int] -> [Int]
mlist' _ _ [] = []
mlist' f x (y:ys) =  applied : mlist' f applied ys
		where
			applied = f x y


{- TESTS -}
-- Positive Tests
positive_lists = 
	[
		[2,5,2,9,3,4,0,2], 
		[3,5,7,9,4,4,7,20], 
		[3,19,7,9,4,2,7,20], 
		[1,2], 
		[1,10,2,2,1,1,0,0], 
		[5,3,7,5,6,2,8,5,1], 
		[1,2,3,4,10,6,7,8,9,10] 
	]

positive_results = 
	[
		(0,3,7), 
		(0,7,17), 
		(5,7,18), 
		(0,1,1), 
		(0,1,9), 
		(5,6,6), 
		(0,9,9) 
	]

positive_zip = zip positive_lists positive_results
print_test_positive = show $ map (\x -> (bestInvestment (fst x), snd x)) $ positive_zip
positive_test_success = all (\x -> (bestInvestment (fst x)) == snd x) $ positive_zip

-- Negative Tests
negative_lists = 
	[
		[2,1], 
		[2,2], 
		[3,3,2,2,1,1,0,0]
	]

print_test_negative = show $ map (\x -> (bestInvestment x, nil)) $ negative_lists
positive_test_negative = all (\x -> (bestInvestment x) == nil) $ negative_lists

-- All Tests Correct?
all_success = positive_test_success && positive_test_negative


initial = [2,5,2,9,3,4,0,2]
kmin = minimumList initial
kmax = maximumList initial
kzip = zip kmin kmax

File: ad/ad_assignments/a9.py

DEBUG = True

def bestInvestment(kurs=[]):
	minimum = minimumList(kurs)
	maximum = maximumList(kurs)
	result = bestInvest(minimum, maximum)

	if DEBUG:
		print "initial list: %s" % str(kurs)
		print "minimum list: %s" % str(minimum)
		print "maximum list: %s" % str(maximum)
		print "min_max list: %s" % str(zip(minimum, maximum))
		print "result: %s" % str(result)

	return result


def bestInvest(minimum, maximum):

	kzip = zip(minimum, maximum)

	def minimumIndex():
		max_diff, min_idx = -1, -1
		for idx, elem in enumerate(kzip):
			possible_max = elem[1] - elem[0]
			if possible_max > max_diff:
				max_diff, min_idx = possible_max, idx
		return min_idx

	def maximumIndex(min_idx, max_val):
		k_from_min = kzip[min_idx:]
		max_offset = min_idx - 1
		for elem in k_from_min:
			if elem != max_val:
				return max_offset
			max_offset += 1
		return max_offset

	min_idx = minimumIndex()
	max_val = kzip[min_idx]
	max_idx = maximumIndex(min_idx, max_val)
	max_diff = max_val[1] - max_val[0]

	if max_diff <= 0:
		return (-1, -1, -1)

	return (min_idx, max_idx, max_diff)


def maximumList(kurs):
	result = kurs[:] # copy kurs
	maximum = kurs[-1] # last element
	idx = len(kurs)-1

	# kurs = [2,5,2,9,3,4,0,2]
	# result = [9, 9, 9, 9, 4, 4, 2, 2]
	while idx >= 0:
		if kurs[idx] > maximum:
			maximum = kurs[idx]

		result[idx] = maximum
		idx = idx-1

	return result


def minimumList(kurs):
	result = []
	minimum = kurs[0]
	
	for elem in kurs:
		if elem < minimum:
			minimum = elem
		result.append(minimum)

	return result



# TEST
def price(k):
	return k[2]

def assertEquPrice(expected, was, msg):
	a = expected
	b = was
	if a == b:
		print msg
		return True
	print "Was expected: %s" % str(a)
	print "Actually was: %s" % str(b)
	raise Exception("False! Error: " + str(msg))

def shouldBeAnOptimalOf(expected, course, err_message):
	assertEquPrice(expected, bestInvestment(course), err_message)

def positive_test():
	print "Positive Test"
	shouldBeAnOptimalOf((0,3,7), [2,5,2,9,3,4,0,2], "")

	shouldBeAnOptimalOf((0,7,17), [3,5,7,9,4,4,7,20], "")

	shouldBeAnOptimalOf((5,7,18), [3,19,7,9,4,2,7,20], "")

	shouldBeAnOptimalOf((0,1,1), [1,2], "")

	shouldBeAnOptimalOf((0,1,9), [1,10,2,2,1,1,0,0], "")
	
	shouldBeAnOptimalOf((5,6,6), [5,3,7,5,6,2,8,5,1], "")

	shouldBeAnOptimalOf((0,9,9), [1,2,3,4,10,6,7,8,9,10], "")

def negative_test():
	print "Negative Test"
	shouldBeAnOptimalOf((-1,-1,-1), [2,1], "")

	shouldBeAnOptimalOf((-1,-1,-1), [2,2], "")

	shouldBeAnOptimalOf((-1,-1,-1), [3,3,2,2,1,1,0,0], "")


positive_test()
negative_test()

print "ALL TESTS SUCCEEDED' SWAG SWAG \u266b YOLOSAWAG \u266b"

